### ISSUE TITLE
Compilation time as a function of input dimensionality

### ISSUE THREAD

**Issue by fountaindive** [jfcrenshaw/pzflow#85]
Hi,

First of all, I would like to say how nice this package is. It has been very nice to use and experiment with using normalising flows for tabular data!

I'm interested in applying normalising flows on datasets that could have O(100) columns so I did a quick timing test and these are my results. The main finding is that the JIT compilation step starts to throw the following error when the number of columns is greater than about 30.

```
2023-03-10 10:40:40.407123: E external/org_tensorflow/tensorflow/compiler/xla/service/slow_operation_alarm.cc:133] The operation took 10m16.840481s

********************************
[Compiling module jit_step] Very slow compile?  If you want to file a bug, run with envvar XLA_FLAGS=--xla_dump_to=/tmp/foo and attach the results.
********************************
```

I have a minimal working example here if you would like to run something. It creates dummy data with 10 rows and an increasingly larger number of columns (features) from 2 to 100. Then I just instantiate a new Flow object and train it for 1 epoch and use that as an estimate of a compilation time. The whole code will take about 1.5 hours to run on my machine and the memory usage maybe got to 1GB or more (that estimate might be quite inaccurate though!)

So my questions are:

 - do you have any experience or guidance using pzflow on datasets with this many columns?
 - can you think of anything to improve the compilation performance for these cases?
 - do you agree that the number of samples (rows) in the dataset shouldn't impact the compilation time? I was thinking that it's just the number of columns that would be important, but I could be wrong! For example do you think the compilation time and memory usage would be significantly larger if I had a dataset with 1e6 rows? (I could just test this myself though!) 

```
import jax

jax.config.update("jax_enable_x64", True)

from pzflow import Flow

import pandas as pd
import numpy as np
import datetime

def build_data(n_samples=10, n_features=2):
    feat_names = [f'x{i}' for i in range(n_features)]
    X = np.random.normal(size=(n_samples, n_features))
    df_X = pd.DataFrame({k:X[:,i] for i, k in enumerate(feat_names)})
    return df_X, feat_names

n_samples_arr = np.array([10]).astype(int)
n_features_arr = np.array([2, 5, 10, 15, 20, 30, 50, 100]).astype(int)
print(f"samples array: {n_samples_arr}")
print(f"features array: {n_features_arr}")


res=[]
for n_f in n_features_arr:
    for n_s in n_samples_arr:
        print(f"working: n_f: {n_f}, n_s: {n_s}")
        df_X, feat_names = build_data(n_s, n_f)
        flow = Flow(data_columns=feat_names)
        start_time = datetime.datetime.now()
        flow.train(df_X, verbose=False, epochs=1)
        end_time = datetime.datetime.now()
        duration = end_time - start_time
        res.append([n_s, n_f, duration])
df_res = pd.DataFrame(res, columns=['n_samples', 'n_features', 'duration'])
```
If you run the above code the output dataframe looks like this

![Screenshot 2023-03-10 at 12 14 39](https://user-images.githubusercontent.com/91950069/224313564-245ca313-dcea-4f97-b6ac-81e7e65a56a7.png)

Thanks for any help!

**Comment by jfcrenshaw** [comment]
Hey Sebastian, 

I'm glad you're finding PZFlow useful! I'm down in Chile observing at the moment, so I don't have time to fully investigate this myself, but I think I know why this is happening.

By default, PZFlow constructs a bijector that has a layer for every column in your data. This will probably get too big if you have a ton of columns. Instead, we should manually construct a bijector with fewer columns.

Basically, add the following code before you construct your flow (this assumes you have 100 columns).

```
# get the min/max for each data column
mins = df_X.min(axis=0)
maxs = df_X.max(axis=0)

# determine how many layers in the RollingSplineCoupling
# and how far to shift the dimension in each layer
# note you want nlayers * shift = number of columns in data
# increasing nlayers will increase expressiveness of flow, but is more expensive
nlayers = 10
shift = 10

bijector = Chain(
    ShiftBounds(mins, maxs, 4.),
    RollingSplineCoupling(
        nlayers=nlayers,
        shift=shift,
)

# now create the flow
flow = Flow(data_columns=feat_names, bijector=bijector)
```

As I explain in the comment above,
- `nlayers` sets the number of layers in the bijector. By default, this is set to the number of columns in your data. More layers = more expressive, but also is more expensive and increases the odds of overfitting
- `shift` sets how far the columns are shifted in each layer. I can explain more about this if you want, but for now, suffice it to say that (nlayers * shift) should equal the number of columns in your data

Let me know if this fixes your problem! If not, I can try to play with this myself early next week

**Comment by fountaindive** [comment]
Hi John,

Many thanks for getting back to me and hope your observing trip goes well!

I'm still a bit of a novice when it comes to normalising flows would you mind explaining the idea behind `shift` or point me to a paper.

It seems your hunch has some merit though. By lowering `nlayers` and compensating with `shift` I can basically set the compilation time. Would it be correct to say that with `nlayers` and `shift` we can trade off potential model complexity with compilation time? (I say potential because I guess it's not necessarily true that a model with more layers will be useful for your data).

If this is correct then I guess this is a pretty good solution to my problem. I just hope that my data can be fit with a model with fewer layers :)

**Comment by fountaindive** [comment]
An additional question. If the data has an awkward number of columns e.g. I have a dataset with 26 columns and wanted to use `nlayers = 5` could I use `shift = 6` which results `nlayers * shift = 30`. What happens to the 4 redundant dimensions?
Would I be correct in thinking that so long as my latent dimension is 26 then I'm using the algorithm correctly?

I need to go and understand how layers and shift work :)

Many thanks!

**Comment by jfcrenshaw** [comment]
I'm glad that helped! If you want to understand the details more, I have a PZFlow paper in review that you can read here: https://github.com/jfcrenshaw/pzflow-paper/raw/main-pdf/ms.pdf

I was just thinking about it, and I actually don't think it's necessary that `nlayers * shift = ncolumns`. That condition was to try to preserve the idea that the first latent dimension gets transformed into the first data dimension, the second latent dimension gets transformed into the second data dimension, etc.

But that's not necessary, and I'm not even sure it makes a lot of sense to think that way anyway...

So yes, with 26 columns, you can use `nlayers = 5` and `shift = 6`. (If this turns out not to work, I would love to know!)

**Comment by fountaindive** [comment]
Ah great I will check out your draft thanks!

Yeah that seems to work well so that's a way to deal with any number of features. One of the toy datasets I tested on had 63 features and I used `nlayers = 5` and `shift = 13` which worked out quite well in terms of speed and accuracy.

I'm currently looking into how to handle categorical features, I saw you have some dequantisation routines but I think they are better suited for ordinal-type variables. I found a couple of papers that propose some methods to handle proper categorical features. If I can get it to work well I would be happy to try and add it to pzflow. But that could be several months down the line tbh.

Many thanks again!

**Comment by jfcrenshaw** [comment]
Glad it's working well! And that would be awesome, thanks!
